### PR TITLE
Add support for TensorRT-RTX

### PR DESCRIPTION
**Pull request by lanluo-nvidia**
# Description

Add initial support for TensorRT-RTX.
The following are the currently identified issues:
RTX team side:
1. mobilenet/efficientnet test failed when input dtype is bfloat16(float16 and float32 is working)
[5439176](https://nvbugswb.nvidia.com/NVBugs5/redir.aspx?url=/5439176)
2. Flux model FP4 and FP8 are not working (fp16 and fp32 is working)
[5400490](https://nvbugswb.nvidia.com/NVBugs5/redir.aspx?url=/5400490)
3. NoneZero has not been supported yet in RTX(will add support in future)
[5407733](https://nvbugswb.nvidia.com/NVBugs5/redir.aspx?url=/5407733)
4. Int8 quantization only support weights quantization, does not support activations quantization yet.
[5402295](https://nvbugswb.nvidia.com/NVBugs5/redir.aspx?url=/5402295)
5. deconv3d is not working in rtx
[5459328](https://nvbugspro.nvidia.com/bug/5459328)
6. cuda graph is not working in rtx
[5481821](https://nvbugspro.nvidia.com/bug/5481821)

Our side:
1. a few test cases failed on strong typing
PR is in Progress:
 https://github.com/pytorch/TensorRT/pull/3751
 https://github.com/pytorch/TensorRT/pull/3749
2. batchnorm constant fold got nan (skip constant fold is able to get the result)
https://github.com/pytorch/TensorRT/issues/3699
3. refit test failed on missing eps weights due to we only capture (bias/weight/running_mean/running_var)
https://github.com/pytorch/TensorRT/issues/3752
4. cuda graph test failed
https://github.com/pytorch/TensorRT/issues/3781
5. torch script test failed
https://github.com/pytorch/TensorRT/issues/3782

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: .github/workflows/build-test-linux-x86_64_rtx.yml
Comment (narendasan) [inline .github/workflows/build-test-linux-x86_64_rtx.yml:1]: can you prefix the RTX name? Like `[RTX] Build and Test Linux x86_64 wheels`

File: .github/workflows/build-test-windows_rtx.yml
Comment (narendasan) [inline .github/workflows/build-test-windows_rtx.yml:1]: Same here

File: core/conversion/conversionctx/ConversionCtx.cpp [hunk 1]
```diff
@@ -31,8 +31,8 @@ std::ostream& operator<<(std::ostream& os, const BuilderSettings& s) {
     if (s.device.device_type == nvinfer1::DeviceType::kDLA) {
     os << "\n    DLACore: " << s.device.dla_core;
     }
-    os << "\n    Engine Capability: " << s.capability                                      \
-       << "\n    Calibrator Created: " << (s.calibrator != nullptr);
+    os << "\n    Engine Capability: " << s.capability;
+      //  << "\n    Calibrator Created: " << (s.calibrator != nullptr);
```
Comment (narendasan) [inline core/conversion/conversionctx/ConversionCtx.cpp]: You can just delete this.

Comment (narendasan) [reply core/conversion/conversionctx/ConversionCtx.cpp]: Also please note in the PR that this will be a breaking change removing the PTQ calibrator

File: core/conversion/conversionctx/ConversionCtx.cpp [hunk 2]
```diff
@@ -59,11 +59,16 @@ ConversionCtx::ConversionCtx(BuilderSettings build_settings)
   for (auto p = settings.enabled_precisions.begin(); p != settings.enabled_precisions.end(); ++p) {
```
Comment (narendasan) [inline core/conversion/conversionctx/ConversionCtx.cpp:58]: Cant we condition this full block? since enabled precisions doesnt matter to RTX? Also the else case of the macro should be a warning that the setting enabled precision does not do anything for TRT RTX

File: core/conversion/conversion.cpp [hunk 3]
```diff
@@ -204,7 +204,7 @@ void AddInputs(ConversionCtx* ctx, c10::ArrayRef<const torch::jit::Value*> input
       "Optimization profile is invalid, please check the input range provided (conversion.AddInputs)");
 
   ctx->cfg->addOptimizationProfile(profile);
-#if NV_TENSORRT_MAJOR > 7 || (NV_TENSORRT_MAJOR == 7 && NV_TENSORRT_MINOR >= 1)
+#ifndef TRT_MAJOR_RTX && (NV_TENSORRT_MAJOR > 7 || (NV_TENSORRT_MAJOR == 7 && NV_TENSORRT_MINOR >= 1))
```
Comment (narendasan) [inline core/conversion/conversion.cpp]: Does this work? like a ifndef + some other conditional?

File: cpp/bin/torchtrtc/main.cpp [hunk 4]
```diff
@@ -334,8 +336,12 @@ int main(int argc, char** argv) {
   if (calibration_cache_file) {
     calibration_cache_file_path = torchtrtc::fileio::resolve_path(args::get(calibration_cache_file));
   }
-
+#ifndef TRT_MAJOR_RTX
```
Comment (narendasan) [inline cpp/bin/torchtrtc/main.cpp]: Same thing here, the else cases should log a warning if the user tries to use the API. Or in this case maybe a error

File: py/torch_tensorrt/dynamo/conversion/impl/quantize.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 import torch
 from torch.fx.experimental.proxy_tensor import unset_fake_temporarily
 from torch.fx.node import Target
+from torch_tensorrt._utils import is_tensorrt_rtx
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/impl/quantize.py]: Use the enabled features system for this instead of a utility function

File: py/torch_tensorrt/fx/converters/acc_ops_converters.py [hunk 6]
```diff
@@ -12,6 +12,7 @@
 import torch
 from torch.fx.immutable_collections import immutable_list
 from torch.fx.node import Argument, Target
+from torch_tensorrt._utils import is_tensorrt_version_supported
```
Comment (narendasan) [inline py/torch_tensorrt/fx/converters/acc_ops_converters.py]: Do we need these FX frontend updates?

File: py/torch_tensorrt/trt_alias.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        libpath = os.path.join(path, name)
+        if os.path.isfile(libpath):
+            return libpath
+
+    raise FileNotFoundError(f"Could not find {name}\n  Search paths: {paths}")
+
+
+# TensorRTProxyModule is a proxy module that allows us to register the tensorrt or tensorrt-rtx package
+# since tensorrt-rtx is the drop-in replacement for tensorrt, we can use the same interface to use tensorrt-rtx
+class TensorRTProxyModule(ModuleType):
```
Comment (narendasan) [inline py/torch_tensorrt/trt_alias.py]: Nice this is cool, Id just rename the file `_TensorRTProxyModule.py`

File: toolchains/ci_workspaces/MODULE.bazel.tmpl [hunk 8]
```diff
@@ -81,6 +81,15 @@ http_archive(
     ],
 )
 
+http_archive(
+    name = "tensorrt_rtx",
```
Comment (narendasan) [inline toolchains/ci_workspaces/MODULE.bazel.tmpl:85]: is there a TRT RTX for ARM or Jetson?

File: pyproject_rtx.toml.temp
Comment (narendasan) [inline pyproject_rtx.toml.temp:1]: I feel like there is a way to handle this without needing to fully replicate the pyproject.toml.

I saw this in vllm `https://github.com/vllm-project/vllm/blob/31f5dc5b2a5da18bc17240c7a67e8770d00901d8/pyproject.toml#L34` that connects to some requiremnts.txt https://github.com/vllm-project/vllm/tree/main/requirements

File: setup.py [hunk 9]
```diff
@@ -115,6 +123,10 @@ def load_dep_info():
     if py_only_env_var == "1":
         PY_ONLY = True
 
+if (use_rtx_env_var := os.environ.get("FORCE_TENSORRT_RTX")) is not None:
```
Comment (narendasan) [inline setup.py]: why not just `USE_TRT_RTX`?

File: .github/scripts/install-tensorrt-rtx.sh [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        if [ -z "$PYTHON_VERSION" ]; then
+            echo "Error: PYTHON_VERSION environment variable is not set or empty. example format: export PYTHON_VERSION=3.11"
+            exit 1
+        fi
+
+        # python version is like 3.11, we need to convert it to cp311
+        CPYTHON_TAG="cp${PYTHON_VERSION//./}"
+        if [[ ${PLATFORM} == win32 ]]; then
+            curl -L https://developer.nvidia.com/downloads/trt/rtx_sdk/secure/1.0/TensorRT-RTX-1.0.0.21.Windows.win10.cuda-12.9.zip -o TensorRT-RTX-1.0.0.21.Windows.win10.cuda-12.9.zip
+            unzip TensorRT-RTX-1.0.0.21.Windows.win10.cuda-12.9.zip
```
Comment (narendasan) [inline .github/scripts/install-tensorrt-rtx.sh]: Might want to templatized the version

File: .github/scripts/install-tensorrt-rtx.sh [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+            rtx_lib_dir=${PWD}/TensorRT-RTX-1.0.0.21/lib
+            export PATH=${rtx_lib_dir}:$PATH
+            echo "PATH: $PATH"
+            if [[ ${install_wheel_or_not} == true ]]; then
+                pip install TensorRT-RTX-1.0.0.21/python/tensorrt_rtx-1.0.0.21-${CPYTHON_TAG}-none-win_amd64.whl
```
Comment (narendasan) [inline .github/scripts/install-tensorrt-rtx.sh]: Same here

File: .github/scripts/install-tensorrt-rtx.sh [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+            fi
+        else
+            curl -L https://developer.nvidia.com/downloads/trt/rtx_sdk/secure/1.0/TensorRT-RTX-1.0.0.21.Linux.x86_64-gnu.cuda-12.9.tar.gz -o TensorRT-RTX-1.0.0.21.Linux.x86_64-gnu.cuda-12.9.tar.gz
```
Comment (narendasan) [inline .github/scripts/install-tensorrt-rtx.sh]: Same here

File: docsrc/getting_started/tensorrt_rtx.rst [hunk 13]
```diff
@@ -0,0 +1,141 @@
+.. _Torch-TensorRT_in_RTX:
```
Comment (narendasan) [inline docsrc/getting_started/tensorrt_rtx.rst]: Make sure this is listed in index.rst

File: docsrc/getting_started/tensorrt_rtx.rst [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+Torch-TensorRT in RTX
```
Comment (narendasan) [inline docsrc/getting_started/tensorrt_rtx.rst]: Call this Torch-TensorRT-RTX or something

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 15]
```diff
@@ -90,11 +89,19 @@ def __init__(
...
-        if compilation_settings.use_explicit_typing:
-            STRONGLY_TYPED = 1 << (int)(
-                trt.NetworkDefinitionCreationFlag.STRONGLY_TYPED
-            )
-            flag |= STRONGLY_TYPED
+        # rtx build, strongly typed is enabled by default, can not set it by builder config
+        if is_tensorrt_rtx():
+            if not compilation_settings.use_explicit_typing:
+                warnings.warn(
+                    "Strongly typed is enabled by default in rtx build,  setting use_explicit_typing to True"
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: Use the full name of the library for the sake of clarity

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 16]
```diff
@@ -90,11 +89,19 @@ def __init__(
         self.builder = trt.Builder(self.logger)
         self._debugger_config = _debugger_config
         flag = 0
-        if compilation_settings.use_explicit_typing:
-            STRONGLY_TYPED = 1 << (int)(
-                trt.NetworkDefinitionCreationFlag.STRONGLY_TYPED
-            )
-            flag |= STRONGLY_TYPED
+        # rtx build, strongly typed is enabled by default, can not set it by builder config
+        if is_tensorrt_rtx():
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: Can we pull this into the enabled features API?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: Or we could pull it into the `is_tensorrt_version_supported` one like an `1.2.0+rtx` version

File: py/torch_tensorrt/dynamo/conversion/impl/quantize.py [hunk 17]
```diff
@@ -77,6 +78,17 @@ def quantize(
             dtype = trt.DataType.FP8
             max_bound = 448
 
+        if (
+            dtype == trt.DataType.INT8
+            and ".input_quantizer" in name
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/impl/quantize.py]: Can we write a validator for this case? (As an aside have we tried quantization where not all of the model is running in TensorRT?)

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 18]
```diff
@@ -3592,25 +3593,29 @@ def aten_ops_full(
...
-) -> Union[TRTTensor, Sequence[TRTTensor]]:
-    return impl.unary.nonzero(
-        ctx,
-        target,
-        SourceIR.ATEN,
-        name,
-        args[0],
+# currently nonzero is not supported for tensorrt_rtx
+# TODO: lan to remove this once rtx team has fixed the bug
+if not ENABLED_FEATURES.tensorrt_rtx:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: Could just stack the decorator here

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: I think
 
```py
@needs_tensorrt_rtx
@dynamo_tensorrt_converter(...)
def aten...
```

Should do what we want

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: Or sorry we should create an inverted decorator for this? Like `not_supported_in_tensorrt_rtx`
